src/functional_simulator.cpp

```diff
@@ -496,4 +507,18 @@ bool FunctionalSimulator::needsRtValue(const Instruction* instr) const {
         default:
             return false;  // Most I-type instructions only use Rs
     }
+}
+
+void FunctionalSimulator::checkProgramCompletion() {
+    if(!halt_pipeline) {
+        return;  // Halt instruction not encountered, program not finished
+    }
+    // If halt instruction was encountered, check if the pipeline is empty
+    for (const auto& stage : pipeline) {
+        if (stage && !stage->isEmpty()) {
+            return;  // Pipeline is not empty, program still running
+        }
+    }
+    // If we reach here, pipeline is empty and halt encountered
+    program_finished = true;  // Set program finished flag
```

Comment:
Instead of adding another control signal, we could have this function return a boolean. I believe this should work, since `checkProgramCompletion` is called at the end of `cycle`, and then `program_finished` is checked immediately at the beginning of `cycle`. So we could change that to `isProgramComplete` called at the beginning of `cycle`.